test/distributed/rpc/test_tensorpipe_agent.py

```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env python3
+import unittest
+
+from torch.testing._internal import dist_utils
+from torch.testing._internal.common_distributed import MultiProcessTestCase
+from torch.testing._internal.common_utils import TEST_WITH_ASAN, run_tests
+from torch.testing._internal.distributed.ddp_under_dist_autograd_test import (
+    TestDdpComparison,
+    TestDdpUnderDistAutograd,
+)
+from torch.testing._internal.distributed.rpc.dist_autograd_test import DistAutogradTest
+from torch.testing._internal.distributed.rpc.dist_optimizer_test import (
+    DistOptimizerTest,
+)
+from torch.testing._internal.distributed.rpc.rpc_test import TensorPipeAgentRpcTest
+from torch.testing._internal.distributed.rpc.tensorpipe_rpc_agent_test_fixture import (
+    TensorPipeRpcAgentTestFixture,
+)
+
+
+# FIXME This is needed to make some functions in dist_init work. Those functions
+# should be moved to methods of the fixture. When that is done, remove this.
+dist_utils.TEST_CONFIG.rpc_backend_name = "TENSORPIPE"
+
+
+@unittest.skipIf(
+    TEST_WITH_ASAN, "Skip ASAN as torch + multiprocessing spawn have known issues"
+)
+class SpawnHelper(TensorPipeRpcAgentTestFixture, MultiProcessTestCase):
+    def setUp(self):
+        super().setUp()
+        self._spawn_processes()
+
+
+class TensorPipeRpcTestWithSpawn(TensorPipeAgentRpcTest, SpawnHelper):
+    pass
+
+
+class TensorPipeDistAutogradTestWithSpawn(DistAutogradTest, SpawnHelper):
+    pass
+
+
+class TensorPipeDistOptimizerTestWithSpawn(DistOptimizerTest, SpawnHelper):
+    pass
+
+
+class TensorPipeTestDdpUnderDistAutograd(TestDdpUnderDistAutograd, SpawnHelper):
+    pass
+
+
+class TensorPipeTestDdpComparison(TestDdpComparison, SpawnHelper):
+    pass
```

Comment:
Since we're doing a major refactor here, should we try to avoid having separate files for tensorpipe tests? The current setup with Thrift and PGA is nice where we have just one test file and appropriate build rules/files enable the appropriate agent.

Can we have a single test file (ex: rpc_test.py, dist_autograd_test.py) that works for all RPC backends? We can have separate build rules for each agent where they set an appropriate environment variable and the test file initializes the appropriate agent based on the environment variable. This ensures that we only modify a single file for new tests and it gets enabled for all backends.

With a separate file for tensorpipe like this, we run the risk of someone adding a new class to `rpc_test.py` and forgetting to add it here.

Reply:
In my opinion we should try to avoid having a single file and then instantiate it several times for different agents (using env vars, or traps, or overrides, ...). In a nutshell, this is because as soon as the logic of the tests becomes too complicated, we have to start testing the tests to make sure they're testing the right thing. And this stack does actually include examples of test logic being incorrect and causing tests to be missed/skipped.

More concretely, though, I don't know if using env vars is practical. First of all, while I am sure there is a way to automatically provide them with Buck, I wouldn't know how to make it work with the OSS test runner (i.e., the `test/run_test.py` script). We'd have to add special logic to that file, incurring in the risk above. Moreover, this would make it harder to run these tests on their own (which is something that is needed particularly in the most inconvenient situations, like when debugging a live failure on CircleCI) as one needs to always remember to pass the right env vars, with the risk of forgetting or mistyping them. Finally, there's also the issue that not all test suites are run on all agents: the faulty agent is an egregious example as it runs on a completely separate set of tests, thus in order to deal with it in a single file with all other agents we'd need a bunch of dedicated logic.

Then there's the question of how to enable/disable tests for a specific agent. You're suggesting decorators but in #40162 I was told to not do that and instead use separate classes, so in #40819 I'm removing decorators entirely. Also, decorators also add further logic that could fail (and it did already fail, very silently, in #40860). So, overall, I'm leaning against decorators.

A few more remarks:
- Note that as part of this stack I am also "duplicating" the PG and Thrift tests which now will use two different entry point script rather than a single one with overrides (see explanation in PR summary for why).
- However, I am also changing the name of the PG entry point to contain `process_group`. This means that there will be no test entry point called "just" `rpc_test`: all entry points will have the name of an agent. So if someone were to add a new test suite, they would be adding it to the `test_process_group_agent.py` file and at that point I hope they will be wondering why they are not adding it to the other files.
- This way of doing it is already much better than what we had before, as now we just have one file per agent, whereas before we had one file per agent, _per suite_ and _per fork/spawn_.

If you'd like, we could go even further and instead of having one file per agent we could have one file total. Inside that file, though, we'd still be explicitly instantiating each test suite several times, once per agent (so we would not have any "automatic" instantiation based on env vars). However, hopefully, this would make it even harder to forget to enroll some agent when adding a new suite.

Reply:
> If you'd like, we could go even further and instead of having one file per agent we could have one file total. Inside that file, though, we'd still be explicitly instantiating each test suite several times, once per agent (so we would not have any "automatic" instantiation based on env vars). However, hopefully, this would make it even harder to forget to enroll some agent when adding a new suite.

Apologize for the delay on getting back to you on this. It would be nice if we can have one file with all the test suites. Although, my point is around mostly having developers to modify a single class/file when they add a new test and not worry about multiple backends. So, we could have a single base class where we add all the tests, but multiple subclasses (for each agent) which actually run them. The per agent classes could be in separate files or the same file, it doesn't matter. I'm wondering if it's possible to have a base class like `TestRPCAll(TensorPipeAgentRpcTest, DistAutogradTest, DistOptimizerTest, ...)` that encompasses all tests? That way each agent just inherits from this base and when we add a new test class, it has to be added here for it to work for any backend and that way we also ensure it's enabled for all backends automatically. Another option would be to have a single master list of test classes we'd like to run and have each per agent class, look up that list and run those tests.

Reply:
I see what you mean and I think it makes sense, but I also see a few "pitfalls" that may make it hard to implement well:
- Merging all the test classes in a single class, by inheriting from them, will cause failing tests to be reported as belonging to that "catch-all" class, thus hiding what test suite they were actually defined in (e.g., we might see `TestRPCAll.test_gpu_simple`, without a way of knowing that it comes from the dist autograd suite). This will make it harder to understand the failure and find the test method.
- We want some agents to run more or fewer tests than others (e.g., TensorPipe and ProcessGroup have a couple of specific tests that should run just on them; Faulty should run just an entirely different set of tests). This breaks the uniformity and requires adding special cases that may hurt readability.
- We have some tests that live in a separate internal-only file (for example all the tests for Thrift) and if we were to have a "master list" we'd need a way to share it between these files.
- Different "modes" (spawn vs fork) need different lists, as for example DdpUnderDistAutogradTest and DdpComparisonTest are only run with spawn but not with fork.

We could address these issues with some extra logic, but as I said I'm against too much logic in tests as getting it wrong may have disastrous consequences (tests that look like they're passing but aren't doing anything).

If you have a concrete proposal of how this approach could be implemented I'd love to hear it, otherwise we could first land this (which is already an improvement on what we had before!) and think about it further.

Reply:
The pitfalls you mentioned above are indeed issues and I was wondering if this modified proposal would work:

1) We maintain a list of all test classes we'd like to run as part of the RPC test_suite. We can maintain this list in `dist_utils.py` or create another file to maintain such common information. These should be all the common test classes that should be run across all agents. Agent specific tests can move to agent specific test files as I've mentioned below. We can maintain a list of tests to skip per agent in the same common location as well. 
2) Have a test file per agent as you proposed, but instead of developers listing out each test class, autogenerate all the test classes from the list in 1).
3) For tests that need to be run only for a particular agent, we create a separate Test class in the agent's test file and add it there.
4) For tests that should be skipped for a particular agent, the autogeneration logic in 2) can take a list of test methods to skip and as part of autogenerating the class, we can skip those methods. A very simple example of what this might look like:

```
from unittest import TestCase

class DistAutogradTest(TestCase):
    def test1(self):
        pass

    def test2(self):
        pass

class DistOptimizerTest(TestCase):
    def test3(self):
        pass

    def test4(self):
        pass

def skip_test_function(self):
    pass

def autogenerate_test(test_prefix, test_classes, tests_to_skip_per_class):
    for test_class in test_classes:
        attrs = {}
        if test_class in tests_to_skip_per_class:
            skip_tests = tests_to_skip_per_class[test_class]
            for skip_test in skip_tests:
                attrs[skip_test] = skip_test_function

        name = test_prefix + test_class.__name__
        print (attrs)
        t = type(name, (test_class,), attrs)
        globals()[name] = t

autogenerate_test('TensorPipe', [DistAutogradTest, DistOptimizerTest], {DistAutogradTest: ["test1"]})
```
5) For internal tests like Thrift, we can have a separate test file for it and just call the autogeneration logic. The Thrift tests can always refer to the master list and the autogeneration API since those would be in OSS.
6) For FaultyProcessGroupAgent, I guess we should just maintain a separate test file all together since those are a completely different test suite.
7) The auto generation logic can also take "spawn"/"fork" tags per test class and generate spawn and fork classes or only one of them if needed.

The idea here is very similar to what you have in this stack, the only major difference is auto generating the classes instead of having users write out `TensorPipeRpcTestWithSpawn` and `TensorPipeDistAutogradTestWithSpawn` manually. This would ensure if we add a new test class, we just add it to the master list in one place and all the required test classes are generated automatically.